blobstore/util: wrap all blobstore impl with prefix

sluongng opened the pull request:
Introduce PrefixBlobstore to wrap arround all existing blobstore
implementations to provide prefix support out-of-the-box.  This should
help us avoid the overhead of having to call util.BlobPath explicitly in
each blobstore implementation.

until.BlobPath is kept exported so that we could mock it during testing.

This may conflict with #3717 and require a merge conflict resolve if
this is merged after.
In which case, author could re-use this comby command for refactoring.

```bash
comby 'util.BlobPath(:[1])' ':[1]' server/backends/blobstore/**/*.go -exclude server/backends/blobstore/util/
```

<!--
Optional: Provide additional description (beyond the PR title).
Description should provide any background or motivation needed for the change, as well
as a high-level overview of the approach taken (if the change is not straightforward).
Detailed rationale for specific sections of the code are probably better off as code comments
so that future readers of that code can benefit.
-->

---

**Version bump**: Patch <!-- Required. Choose from: Major, Minor, Patch, None -->

<!-- See https://semver.org/#semantic-versioning-specification-semver. Summary:
* Major: Breaking change that causes existing functionality to not work as expected.
* Minor: Non-breaking change that adds functionality (examples: new feature; new API options)
* Patch: Non-breaking change that fixes an issue, improves performance, or refactors
         code.
* None:  Changed files are not included in releases (tests, docs, development setup,
         production configs)
-->

<!-- Optional:
**Related issues**: Fixes #1, Unblocks #2 ...
-->

bduffany commented on server/backends/blobstore/util/util.go:
```diff
@@ -67,6 +69,30 @@ var BlobPath = func(blobName string) string {
+	interfaces.Blobstore
+}
+
+func NewPrefixBlobstore(b interfaces.Blobstore) *prefixBlobstore {
```
nit: it might be worth having a method `WithPrefix(b)` that just returns `b` if prefix is empty, else returns `NewPrefixBlobstore(b)`. That way we avoid the `prefixBlobstore` overhead for the common case of not setting a prefix.

bduffany reviewed (approved):
this LGTM if you want to pull in Zoey's `Writer()` method and merge :+1:

sluongng commented:
I made some updates to this PR as I rebased on top of https://github.com/buildbuddy-io/buildbuddy/pull/3717

Added test cases for the `Writer()` method.

bduffany commented on server/backends/blobstore/util/util.go:
```diff
@@ -71,8 +71,47 @@ func Compress(in []byte) ([]byte, error) {
+	return &prefixBlobstore{b, prefix}
+}
+
+func (p *prefixBlobstore) blobPath(blobName string) string {
+	return filepath.Join(p.prefix, blobName)
+}
```
since this func exists I think we can delete `util.BlobPath` now? Do you want to do that in this PR?

sluongng replied:
I did remove util.BlobPath in this PR 🤔 , do you mean removing `prefixBlobstore.blobPath()` in favor of using `filepath.Join()` directly?

bduffany replied:
woops, ignore me - I missed the diff :sweat_smile:
